File: src/main/java/com/gmail/borlandlp/minigamesdtools/arena/ArenaBase.java
package com.gmail.borlandlp.minigamesdtools.arena;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunkLoaderController;
import com.gmail.borlandlp.minigamesdtools.arena.customhandlers.HandlersController;
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyInitializedException;
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyStartedException;
import com.gmail.borlandlp.minigamesdtools.arena.gui.hotbar.HotbarController;
import com.gmail.borlandlp.minigamesdtools.arena.gui.providers.GUIController;
import com.gmail.borlandlp.minigamesdtools.conditions.ConditionsChain;
import com.gmail.borlandlp.minigamesdtools.arena.phasecomponent.PhaseComponentController;
import com.gmail.borlandlp.minigamesdtools.arena.scenario.ScenarioChainController;
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamController;
import com.gmail.borlandlp.minigamesdtools.events.ArenaGameEndedEvent;
import org.bukkit.Bukkit;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.time.Instant;

public class ArenaBase {
    protected String name;
    protected boolean enabled;
    protected ArenaBase.STATE state = STATE.EMPTY;
    protected int currentRound = 1;
    protected String gameId;

    protected long game_unix_startTime;
    protected long round_unix_startTime;
    protected BukkitTask schedulerTask;
    protected BukkitTask countdownTask;
    protected long countdown_startTime;

    protected long delayedCountdown_startTime;
    protected long delayedCountdown_waitTime = 5;
    protected BukkitTask delayedCountdown_task;

    protected EventAnnouncer eventAnnouncer = new EventAnnouncer();
    protected PhaseComponentController phaseComponentController = new PhaseComponentController();
    protected HandlersController handlersController = new HandlersController();

    protected GUIController guiController;
    protected ScenarioChainController scenarioChainController;
    protected TeamController teamController;
    protected ChunkLoaderController chunkLoaderController;
    protected HotbarController hotbarController;
    protected GameRules gameRules;
    protected ConditionsChain joinConditionsChain;

    private boolean initialized = false;

    public void forceDisable() {
        this.gameEnded(true);
        this.enabled = false;
    }

    public void initializeComponents() throws ArenaAlreadyInitializedException {
        if(this.initialized) {
            throw new ArenaAlreadyInitializedException();
        }

        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.INIT);
        this.initialized = true;
    }

    public void startArena() throws ArenaAlreadyStartedException {
        if(this.getState() != STATE.EMPTY && this.getState() != STATE.DELAYED_START) {
            throw new ArenaAlreadyStartedException();
        }

        ArenaBase arenaObj = this;

        arenaObj.beforeGameStarting();
        arenaObj.beforeRoundStarting();

        this.schedulerTask = new BukkitRunnable() {
            public void run() {
                // countdown pause
                if(arenaObj.getState() == STATE.COUNTDOWN || arenaObj.getState() == STATE.PAUSED) {
                    return;
                }

                arenaObj.update();

                if(arenaObj.getScenarioChainController().hasSignalGameEnding()) {
                    arenaObj.gameEnded(false);
                } else if(arenaObj.getScenarioChainController().hasSignalRoundEnding()) {
                    arenaObj.roundEnded();
                    arenaObj.setCurrentRound(arenaObj.getCurrentRound() + 1);
                    arenaObj.beforeRoundStarting();
                }
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 10, 10);
    }

    protected void delayedStartArena() {
        ArenaBase arenaBaseObject = this;
        this.state = STATE.DELAYED_START;
        arenaBaseObject.delayedCountdown_startTime = Instant.now().getEpochSecond();
        Debug.print(Debug.LEVEL.NOTICE, "Delayed start arena " + this.getName());

        this.delayedCountdown_task = new BukkitRunnable() {
            public void run() {
                int inc = (int)(Instant.now().getEpochSecond() - (int) arenaBaseObject.delayedCountdown_startTime);
                if(inc > arenaBaseObject.delayedCountdown_waitTime) {
                    try {
                        arenaBaseObject.startArena();
                    } catch (ArenaAlreadyStartedException e) {
                        e.printStackTrace();
                    }
                    arenaBaseObject.delayedCountdown_task.cancel();
                }
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 0, 20);
    }

    public void cancelDelayedStartArena() {
        if(this.delayedCountdown_task == null || this.delayedCountdown_task.isCancelled()) {
            return;
        }
        Debug.print(Debug.LEVEL.NOTICE, "Cancel delayed start arena " + this.getName());

        this.delayedCountdown_task.cancel();
    }

    protected void beforeGameStarting() {
        this.game_unix_startTime = Instant.now().getEpochSecond();
        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.GAME_STARTING);
    }

    protected void beforeRoundStarting() {
        ArenaBase arenaBaseObject = this;

        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.ROUND_STARTING);

        this.state = STATE.COUNTDOWN;
        arenaBaseObject.countdown_startTime = Instant.now().getEpochSecond();
        this.countdownTask = new BukkitRunnable() {
            public void run() {
                int inc = (int)(Instant.now().getEpochSecond() - (int) arenaBaseObject.countdown_startTime);
                if(inc > arenaBaseObject.getGameRules().beforeRoundStartingWaitDuration) {
                    arenaBaseObject.state = STATE.IN_PROGRESS;
                    arenaBaseObject.round_unix_startTime = Instant.now().getEpochSecond();
                    arenaBaseObject.countdownTask.cancel();
                }
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 0, 20);
    }

    protected void update() {
        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.UPDATE);
    }

    protected void roundEnded() {
        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.ROUND_ENDING);
    }

    protected void gameEnded(boolean force) {
        this.setState(STATE.ENDED);
        this.game_unix_startTime = 0;

        if(this.schedulerTask != null) this.schedulerTask.cancel();
        if(this.countdownTask != null) this.countdownTask.cancel();

        if(this.getState() == STATE.PAUSED) {
            return;
        }

        try {
            Bukkit.getServer().getPluginManager().callEvent(new ArenaGameEndedEvent(this, Result.NORMAL));
        } catch (Exception ex) {
            ex.printStackTrace();
        }

        this.getPhaseComponentController().announceNewPhase(PhaseComponentController.ArenaPhase.GAME_ENDING);
    }

    public HotbarController getHotbarController() {
        return hotbarController;
    }

    public ChunkLoaderController getChunkLoaderController() {
        return chunkLoaderController;
    }

    public EventAnnouncer getEventAnnouncer() {
        return eventAnnouncer;
    }

    public String getGameId() {
        return gameId;
    }

    public ScenarioChainController getScenarioChainController() {
        return scenarioChainController;
    }

    public GameRules getGameRules() {
        return gameRules;
    }

    public GUIController getGuiController() {
        return guiController;
    }

    public long getGameStartTime() {
        return this.game_unix_startTime;
    }

    public long getRoundStartTime() {
        return round_unix_startTime;
    }

    public int getTimeLeft() {
        return (int)(this.getGameRules().roundTime - (Instant.now().getEpochSecond() - this.getRoundStartTime()));
    }

    public int getCurrentRound() {
        return this.currentRound;
    }

    public void setCurrentRound(int amount) {
        this.currentRound = amount;
    }

    public boolean isEnabled() {
        return this.enabled;
    }

    public void setEnabled(boolean mode) {
        this.enabled = mode;
    }

    public ArenaBase.STATE getState() {
        return this.state;
    }

    public void setState(ArenaBase.STATE state) { this.state = state; }

    public TeamController getTeamController() { return this.teamController; }

    public String getName() { return this.name; }

    public PhaseComponentController getPhaseComponentController() {
        return phaseComponentController;
    }

    public HandlersController getHandlersController() {
        return handlersController;
    }

    public ConditionsChain getJoinConditionsChain() {
        return joinConditionsChain;
    }

    public enum STATE {
        DELAYED_START,
        IN_PROGRESS,
        COUNTDOWN,
        PAUSED,
        EMPTY,
        ENDED
    }

    public static Builder newBuilder() {
        return new ArenaBase().new Builder();
    }

    public class Builder {
        private Builder() {

        }

        public void setJoinConditionsChain(ConditionsChain joinConditionsChain) {
            ArenaBase.this.joinConditionsChain = joinConditionsChain;
        }

        public void setTeamController(TeamController t) {
            ArenaBase.this.teamController = t;
        }

        public void setGameId(String gameId) {
            ArenaBase.this.gameId = gameId;
        }

        public void setName(String name) {
            ArenaBase.this.name = name;
        }

        public void setHotbarController(HotbarController hotbarController) {
            ArenaBase.this.hotbarController = hotbarController;
        }

        public void setChunkLoaderController(ChunkLoaderController chunkLoaderController) {
            ArenaBase.this.chunkLoaderController = chunkLoaderController;
        }

        public void setScenarioChainController(ScenarioChainController scenarioChainController) {
            ArenaBase.this.scenarioChainController = scenarioChainController;
        }

        public void setGuiController(GUIController guiController) {
            ArenaBase.this.guiController = guiController;
        }

        public void setGameRules(GameRules rules) {
            ArenaBase.this.gameRules = rules;
        }

        public ArenaBase getArena() {
            return ArenaBase.this;
        }

        public ArenaBase build() {
            return ArenaBase.this;
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/arena/ArenaBase.kt
package com.gmail.borlandlp.minigamesdtools.arena

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.Debug.print
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools.Companion.instance
import com.gmail.borlandlp.minigamesdtools.arena.chunkloader.ChunkLoaderController
import com.gmail.borlandlp.minigamesdtools.arena.customhandlers.HandlersController
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyInitializedException
import com.gmail.borlandlp.minigamesdtools.arena.exceptions.ArenaAlreadyStartedException
import com.gmail.borlandlp.minigamesdtools.arena.gui.hotbar.HotbarController
import com.gmail.borlandlp.minigamesdtools.arena.gui.providers.GUIController
import com.gmail.borlandlp.minigamesdtools.arena.phasecomponent.PhaseComponentController
import com.gmail.borlandlp.minigamesdtools.arena.scenario.ScenarioChainController
import com.gmail.borlandlp.minigamesdtools.arena.team.TeamController
import com.gmail.borlandlp.minigamesdtools.conditions.ConditionsChain
import com.gmail.borlandlp.minigamesdtools.events.ArenaGameEndedEvent
import org.bukkit.Bukkit
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask
import java.time.Instant

class ArenaBase {
    var name: String? = null
        protected set
    var isEnabled = false
    var state =
        STATE.EMPTY
    var currentRound = 1
    var gameId: String? = null
        protected set
    var gameStartTime: Long = 0
        protected set
    var roundStartTime: Long = 0
        protected set
    protected var schedulerTask: BukkitTask? = null
    protected var countdownTask: BukkitTask? = null
    protected var countdown_startTime: Long = 0
    protected var delayedCountdown_startTime: Long = 0
    protected var delayedCountdown_waitTime: Long = 5
    protected var delayedCountdown_task: BukkitTask? = null
    var eventAnnouncer = EventAnnouncer()
        protected set
    var phaseComponentController = PhaseComponentController()
        protected set
    var handlersController = HandlersController()
        protected set
    var guiController: GUIController? = null
        protected set
    var scenarioChainController: ScenarioChainController? = null
        protected set
    var teamController: TeamController? = null
        protected set
    var chunkLoaderController: ChunkLoaderController? = null
        protected set
    var hotbarController: HotbarController? = null
        protected set
    var gameRules: GameRules? = null
        protected set
    var joinConditionsChain: ConditionsChain? = null
        protected set
    private var initialized = false
    fun forceDisable() {
        gameEnded(true)
        isEnabled = false
    }

    @Throws(ArenaAlreadyInitializedException::class)
    fun initializeComponents() {
        if (initialized) {
            throw ArenaAlreadyInitializedException()
        }
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.INIT)
        initialized = true
    }

    @Throws(ArenaAlreadyStartedException::class)
    fun startArena() {
        if (state != STATE.EMPTY && state != STATE.DELAYED_START) {
            throw ArenaAlreadyStartedException()
        }
        val arenaObj = this
        arenaObj.beforeGameStarting()
        arenaObj.beforeRoundStarting()
        schedulerTask = object : BukkitRunnable() {
            override fun run() { // countdown pause
                if (arenaObj.state == STATE.COUNTDOWN || arenaObj.state == STATE.PAUSED) {
                    return
                }
                arenaObj.update()
                if (arenaObj.scenarioChainController!!.hasSignalGameEnding()) {
                    arenaObj.gameEnded(false)
                } else if (arenaObj.scenarioChainController!!.hasSignalRoundEnding()) {
                    arenaObj.roundEnded()
                    arenaObj.currentRound = arenaObj.currentRound + 1
                    arenaObj.beforeRoundStarting()
                }
            }
        }.runTaskTimer(instance, 10, 10)
    }

    fun delayedStartArena() {
        val arenaBaseObject = this
        state = STATE.DELAYED_START
        arenaBaseObject.delayedCountdown_startTime = Instant.now().epochSecond
        print(
            Debug.LEVEL.NOTICE,
            "Delayed start arena " + name
        )
        delayedCountdown_task = object : BukkitRunnable() {
            override fun run() {
                val inc =
                    (Instant.now().epochSecond - arenaBaseObject.delayedCountdown_startTime.toInt()).toInt()
                if (inc > arenaBaseObject.delayedCountdown_waitTime) {
                    try {
                        arenaBaseObject.startArena()
                    } catch (e: ArenaAlreadyStartedException) {
                        e.printStackTrace()
                    }
                    arenaBaseObject.delayedCountdown_task!!.cancel()
                }
            }
        }.runTaskTimer(instance, 0, 20)
    }

    fun cancelDelayedStartArena() {
        if (delayedCountdown_task == null || delayedCountdown_task!!.isCancelled) {
            return
        }
        print(
            Debug.LEVEL.NOTICE,
            "Cancel delayed start arena $name"
        )
        delayedCountdown_task!!.cancel()
    }

    protected fun beforeGameStarting() {
        gameStartTime = Instant.now().epochSecond
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.GAME_STARTING)
    }

    protected fun beforeRoundStarting() {
        val arenaBaseObject = this
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.ROUND_STARTING)
        state = STATE.COUNTDOWN
        arenaBaseObject.countdown_startTime = Instant.now().epochSecond
        countdownTask = object : BukkitRunnable() {
            override fun run() {
                val inc = (Instant.now().epochSecond - arenaBaseObject.countdown_startTime.toInt()).toInt()
                if (inc > arenaBaseObject.gameRules!!.beforeRoundStartingWaitDuration) {
                    arenaBaseObject.state = STATE.IN_PROGRESS
                    arenaBaseObject.roundStartTime = Instant.now().epochSecond
                    arenaBaseObject.countdownTask!!.cancel()
                }
            }
        }.runTaskTimer(instance, 0, 20)
    }

    protected fun update() {
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.UPDATE)
    }

    protected fun roundEnded() {
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.ROUND_ENDING)
    }

    fun gameEnded(force: Boolean) {
        state = STATE.ENDED
        gameStartTime = 0
        if (schedulerTask != null) schedulerTask!!.cancel()
        if (countdownTask != null) countdownTask!!.cancel()
        if (state == STATE.PAUSED) {
            return
        }
        try {
            Bukkit.getServer().pluginManager
                .callEvent(ArenaGameEndedEvent(this, Result.NORMAL))
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
        phaseComponentController.announceNewPhase(PhaseComponentController.ArenaPhase.GAME_ENDING)
    }

    val timeLeft: Int
        get() = (gameRules!!.roundTime - (Instant.now().epochSecond - roundStartTime)).toInt()

    enum class STATE {
        DELAYED_START, IN_PROGRESS, COUNTDOWN, PAUSED, EMPTY, ENDED
    }

    inner class Builder {
        fun setJoinConditionsChain(joinConditionsChain: ConditionsChain?) {
            this@ArenaBase.joinConditionsChain = joinConditionsChain
        }

        fun setTeamController(t: TeamController?) {
            teamController = t
        }

        fun setGameId(gameId: String?) {
            this@ArenaBase.gameId = gameId
        }

        fun setName(name: String?) {
            this@ArenaBase.name = name
        }

        fun setHotbarController(hotbarController: HotbarController?) {
            this@ArenaBase.hotbarController = hotbarController
        }

        fun setChunkLoaderController(chunkLoaderController: ChunkLoaderController?) {
            this@ArenaBase.chunkLoaderController = chunkLoaderController
        }

        fun setScenarioChainController(scenarioChainController: ScenarioChainController?) {
            this@ArenaBase.scenarioChainController = scenarioChainController
        }

        fun setGuiController(guiController: GUIController?) {
            this@ArenaBase.guiController = guiController
        }

        fun setGameRules(rules: GameRules?) {
            gameRules = rules
        }

        val arena: ArenaBase
            get() = this@ArenaBase

        fun build(): ArenaBase {
            return this@ArenaBase
        }
    }

    companion object {
        fun newBuilder(): Builder {
            return ArenaBase().Builder()
        }
    }
}